### PR TITLE
Restructured the KDF definition to more cleanly align with SP800-56Cr2

### PR DESCRIPTION
**Pull request by ounsworth**
My co-worker Serge Mister found that the interaction between this document and the KDF defined in SP 800-56Cr2 were hard to follow. In particular, we were presenting together parameters of our combiner, parameters of -56Cr2 (such as `counter = 0x00000001`), and parameters of KMAC in a way that made it difficult to tell which parameters belong to which layer.

I have attempted to separate these more cleanly and simply use -56Cr2 as a black-box `KDF(Z, OtherInput)`.

### CODE REVIEW

File: draft-ounsworth-cfrg-kem-combiners.mkd [hunk 1]
```diff
@@ -258,35 +261,27 @@ The need for a KEM-based authenticated key establishment arises, for example, wh
...
 {: #tab-kemCombiner title="general KEM combiner construction" }
 
 where:
 
-- `KDF` represents a suitable choice of a cryptographic key derivation function,
+- `KDF` represents a suitable choice of a cryptographic key derivation function as defined in {{sec-instantiation}}.
 - `k_i` represent the constant-length input keys and is discussed in {{sec-k_i-construction}},
 - `fixedInfo` is some protocol-specific KDF binding,
-- `counter` parameter is instantiation-specific and is discussed in {{sec-instantiation}}.
-- `outputBits` determines the length of the output keying material,
```
Comment (sergemister) [inline draft-ounsworth-cfrg-kem-combiners.mkd:289]: Line 411 includes the outputBits parameter, but it has been removed here.  My guess is that you would want to include it.  I see in SP800-56C that they define KDM( Z, OtherInput), where OtherInput includes the output length.

File: draft-ounsworth-cfrg-kem-combiners.mkd [hunk 2]
```diff
@@ -396,42 +391,49 @@ This is a non-comprehensive list, further information can be found in paragraph
...
-The KDF must be instantiated with cryptographically-secure choices for `KDF`. The following are RECOMMENDED Keccak-based instatiations, but other choices MAY be made for example to allow for future cryptographic agility. A protocol using a different instantiation MUST justify that it will behave as a split-key PRF, as required in {{GHP18}}.
+The KDF must be instantiated with cryptographically-secure choices for `KDF`. The following are RECOMMENDED KDF constructions based on NIST SP800-56Cr2 {{SP800-56C}} with Keccak-based instatiations, but other choices MAY be made for example to allow for future cryptographic agility. A protocol using a different instantiation MUST justify that it will behave as a split-key PRF, as required in {{GHP18}}.
 
-Each instance defines a function to be used as `KDF`, a `hashSize` to determine parameter size, and optionally a `counter`:
-
-1. `KDF = KMAC128`, with `hashSize = 128 bit`.
-1. `KDF = KMAC256`, with `hashSize = 256 bit`.
-1. `KDF = SHA3-256`, with `hashSize = 256 bit`.
-1. `KDF = SHA3-512`, with `hashSize = 512 bit`.
+| Instantiation number | KDF construction | ??       | hashSize | outputBits |
```
Comment (sergemister) [inline draft-ounsworth-cfrg-kem-combiners.mkd]: I think the ?? should be "Auxiliary Function", taken from 4.1 of SP800-56Cr2.

File: draft-ounsworth-cfrg-kem-combiners.mkd [hunk 3]
```diff
@@ -396,42 +391,49 @@ This is a non-comprehensive list, further information can be found in paragraph
...
 
-Each instance defines a function to be used as `KDF`, a `hashSize` to determine parameter size, and optionally a `counter`:
-
-1. `KDF = KMAC128`, with `hashSize = 128 bit`.
-1. `KDF = KMAC256`, with `hashSize = 256 bit`.
-1. `KDF = SHA3-256`, with `hashSize = 256 bit`.
-1. `KDF = SHA3-512`, with `hashSize = 512 bit`.
+| Instantiation number | KDF construction | H(x)     | hashSize | outputBits |
+| ---                  | ---              | ---      | ---      | ---        |
+| 1                    | {#sec-kmac-kdf}  | KMAC128  | 128 bit  | Variable   |
```
Comment (sergemister) [inline draft-ounsworth-cfrg-kem-combiners.mkd]: How about "Option 3 with KMAC128" instead of just "KMAC128"?  And similar for KMAC256?

File: draft-ounsworth-cfrg-kem-combiners.mkd [hunk 4]
```diff
@@ -396,42 +391,49 @@ This is a non-comprehensive list, further information can be found in paragraph
...
+This section defines a `KDF(Z, OtherInput, outputBits)` as per {{SP800-56C}} section 4.1 Option 3 using the KMAC primitive as specified in NIST SP 800-185 {{SP800-185}}.
+To instantiate `KMAC#`:
 
-* The context `S` MUST be the utf-8 string "KDF".
-* The key `K` MUST be a context-specific string of at least `hashSize` bits, and it MAY be used as an additional option to perform context separation, in scenarios where `fixedInfo` is not sufficient.
-* The parameter `counter` MUST be the fixed value `0x00000001`, corresponding to a 32-bit binary big-endian representation of the number zero.
-    * This aligns with {{SP800-56C}} where `counter` is initialized to `0x00000000` is Section 4.1 Step 3, but is first incremented in 6.1 before being used in a hash in 6.2.
+KMAC is defined in NIST SP 800-185 [SP800-185]. The `KMAC#(K, X, L, S)` parameters MUST be instantiated as follows:
+
+* `KMAC#` : either `KMAC128` or `KMAC256` is per {{tab-kdfs}}.
```
Comment (sergemister) [inline draft-ounsworth-cfrg-kem-combiners.mkd]: "as per", instead of "is per"?

File: draft-ounsworth-cfrg-kem-combiners.mkd [hunk 5]
```diff
@@ -396,42 +391,49 @@ This is a non-comprehensive list, further information can be found in paragraph
...
+* `K`: a context-specific string of at least `hashSize` bits, and it MAY be used as an additional option to perform context separation, in scenarios where `OtherInput` is not sufficient.
```
Comment (sergemister) [inline draft-ounsworth-cfrg-kem-combiners.mkd]: Where does the minimum length of the string come from?

File: draft-ounsworth-cfrg-kem-combiners.mkd [hunk 6]
```diff
@@ -396,42 +391,49 @@ This is a non-comprehensive list, further information can be found in paragraph
...
+* `X`: the message input to `KDF()`, as defined above.
+* `L`: integer representation of `outputBits`.
+* `S`: as specified in {{SP800-56C}}, it is the 8-bit ASCII string string "KDF".
+
+
+Since we are setting `L = outputBits`, Step 1 of the process in {{SP800-56C}} spction 4.1 will always be `reps = ciel(L / H_outputBits) = 1`, and since the input strings defined in {{sec-kem-combiner}} are fixed-length and and well below the max input size of KMAC, we can also skip step 4.
```
Comment (sergemister) [inline draft-ounsworth-cfrg-kem-combiners.mkd]: Typo - "spction" and "ciel", and "and and"

Comment (sergemister) [inline draft-ounsworth-cfrg-kem-combiners.mkd]: Re "well below the max input size of KMAC", I don't think there is a max input size; SP800-56Cr2 says "if Option
3 is chosen for the implementation of H, then there is no specification-imposed upper bound
on max_H_inputBits"

File: draft-ounsworth-cfrg-kem-combiners.mkd [hunk 7]
```diff
@@ -396,42 +391,49 @@ This is a non-comprehensive list, further information can be found in paragraph
...
+## Hash-and-counter based construction {#sec-hash-kdf}
 
-* the `counter` MUST be initialized with the value `0x00000001`.
-* The hash is performed over the string defined in {{sec-kem-combiner}}, and repeated `ceil(outputBits/hashSize)` times. For each iteration the `counter` MUST be increased by `0x01`.
-* The strings are concatenated ordered by `counter`.
-* The leftmost `outputBits` are returned as `ss`.
+This section defines a `KDF(Z, OtherInput)` as per {{SP800-56C}} section 4.1 Option 1 using the SHA3 primitive as specified in NIST FIPS 202 {{FIPS202}}.
 
-An implementation MUST NOT overflow and reuse the `counter` and an error MUST be returned when producing more than 2^32 consecutive hashes.
+This KDF construction requires the full KDF construction as specified {{SP800-56C}} section 4.1. No additional parameters are required.
```
Comment (sergemister) [inline draft-ounsworth-cfrg-kem-combiners.mkd]: "as specified" missing "in"

File: draft-ounsworth-cfrg-kem-combiners.mkd [hunk 8]
```diff
@@ -341,9 +336,8 @@ The ciphertext precedes the secret share, as memory-constrained devices can writ
 For a two-KEM instantiation, the construction is
```
Comment (fluppe2) [inline draft-ounsworth-cfrg-kem-combiners.mkd]:
```suggestion
For a two-KEM instantiation with length encoding, the construction is
```

File: draft-ounsworth-cfrg-kem-combiners.mkd (hunk 2 above)
Comment (ounsworth) [reply draft-ounsworth-cfrg-kem-combiners.mkd]: Changed.

File: draft-ounsworth-cfrg-kem-combiners.mkd (hunk 3 above)
Comment (ounsworth) [reply draft-ounsworth-cfrg-kem-combiners.mkd]: Changed.

File: draft-ounsworth-cfrg-kem-combiners.mkd (hunk 4 above)
Comment (ounsworth) [reply draft-ounsworth-cfrg-kem-combiners.mkd]: Changed.

File: draft-ounsworth-cfrg-kem-combiners.mkd (hunk 5 above)
Comment (ounsworth) [reply draft-ounsworth-cfrg-kem-combiners.mkd]: Oops.

File: draft-ounsworth-cfrg-kem-combiners.mkd (hunk 6 above)
Comment (ounsworth) [reply draft-ounsworth-cfrg-kem-combiners.mkd]: Changed.

Comment (ounsworth) [reply draft-ounsworth-cfrg-kem-combiners.mkd]: Good catch.

File: draft-ounsworth-cfrg-kem-combiners.mkd (hunk 7 above)
Comment (ounsworth) [reply draft-ounsworth-cfrg-kem-combiners.mkd]: Changed.

File: draft-ounsworth-cfrg-kem-combiners.mkd (hunk 1 above)
Comment (ounsworth) [reply draft-ounsworth-cfrg-kem-combiners.mkd:289]: Good catch.

**Review by fluppe2 — APPROVED** [review]
Looks good to me

**Review by wussler — APPROVED** [review]
